File: app/src/main/java/org/bubbble/listlabs/java/ProductAdapter.java
package org.bubbble.listlabs.java;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import org.bubbble.listlabs.databinding.ItemProductBinding;
import org.bubbble.listlabs.databinding.ItemProductCategoryBinding;

/**
 * @author Andrew
 * @date 2021/02/08 12:14
 */
public class ProductAdapter extends ListAdapter<ProductItem, RecyclerView.ViewHolder> {

    protected ProductAdapter() {
        super(new ProductDiffCallback());
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());

        switch (viewType) {
            case 0: return new CategoryViewHolder(
                    ItemProductCategoryBinding.inflate(layoutInflater, parent, false)
            );

            case 1: return new ProductViewHolder(
                    ItemProductBinding.inflate(layoutInflater, parent, false)
            );

            default: throw new IllegalArgumentException("Invalid viewType");
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof CategoryViewHolder) {
            ((CategoryViewHolder) holder).onBind(getItem(position));
        }

        if (holder instanceof ProductViewHolder) {
            ((ProductViewHolder) holder).onBind(getItem(position));
        }
    }

    @Override
    public int getItemViewType(int position) {
        switch (getItem(position).getType()) {
            case TITLE: return 0;
            case PRODUCT: return 1;
            default: return -1;
        }
    }
}

class CategoryViewHolder extends RecyclerView.ViewHolder {

    private final ItemProductCategoryBinding binding;

    public CategoryViewHolder(@NonNull ItemProductCategoryBinding binding) {
        super(binding.getRoot());
        this.binding = binding;
    }

    void onBind(ProductItem data) {
        binding.title.setText(data.getCategoryTitle());
    }
}


class ProductViewHolder extends RecyclerView.ViewHolder {

    private final ItemProductBinding binding;

    public ProductViewHolder(@NonNull ItemProductBinding binding) {
        super(binding.getRoot());
        this.binding = binding;
    }

    void onBind(ProductItem data) {
        Glide.with(binding.productImage).load(data.getProductImage()).into(binding.productImage);
        binding.productName.setText(data.getProductName());
    }
}

class ProductDiffCallback extends DiffUtil.ItemCallback<ProductItem> {

    @Override
    public boolean areItemsTheSame(@NonNull ProductItem oldItem, @NonNull ProductItem newItem) {
        return oldItem == newItem;
    }

    @Override
    public boolean areContentsTheSame(@NonNull ProductItem oldItem, @NonNull ProductItem newItem) {
        return oldItem.getProductName().equals(newItem.getProductName())
                && oldItem.getProductImage() == ( newItem.getProductImage())
                && oldItem.getCategoryTitle().equals(newItem.getCategoryTitle())
                && oldItem.getType().equals( newItem.getType());
    }
}

File: app/src/main/java/org/bubbble/listlabs/kotlin/ProductAdapter.kt
package org.bubbble.listlabs.kotlin

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import org.bubbble.listlabs.R
import org.bubbble.listlabs.databinding.ItemProductBinding
import org.bubbble.listlabs.databinding.ItemProductCategoryBinding


/**
 * @author Andrew
 * @date 2021/02/07 13:10
 */
internal class ProductAdapter : ListAdapter<ProductItem, ViewHolder>(ProductDiffCallback) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)

        return when (viewType) {
            R.layout.item_product_category -> ViewHolder.CategoryViewHolder(
                    ItemProductCategoryBinding.inflate(layoutInflater, parent, false))

            R.layout.item_product -> ViewHolder.ProductViewHolder(
                    ItemProductBinding.inflate(layoutInflater, parent, false))

            else -> throw IllegalArgumentException("Invalid viewType")
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (holder is ViewHolder.CategoryViewHolder) {
            holder.onBind(getItem(position))
        }
        if (holder is ViewHolder.ProductViewHolder) {
            holder.onBind(getItem(position))
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when (getItem(position).type) {
            ItemCategory.TITLE -> R.layout.item_product_category
            ItemCategory.PRODUCT -> R.layout.item_product
        }
    }
}

internal sealed class ViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {

    class CategoryViewHolder(private val binding: ItemProductCategoryBinding) : ViewHolder(binding.root) {
        fun onBind(data: ProductItem) {
            binding.title.text = data.categoryTitle
        }
    }

    class ProductViewHolder(private val binding: ItemProductBinding) : ViewHolder(binding.root) {
        fun onBind(data: ProductItem) {
            Glide.with(binding.productImage).load(data.productImage).into(binding.productImage)
            binding.productName.text = data.productName
        }
    }
}

object ProductDiffCallback : DiffUtil.ItemCallback<ProductItem>() {
    override fun areItemsTheSame(oldItem: ProductItem, newItem: ProductItem): Boolean {
        return oldItem != newItem
    }

    override fun areContentsTheSame(oldItem: ProductItem, newItem: ProductItem): Boolean {
        return oldItem != newItem
    }
}